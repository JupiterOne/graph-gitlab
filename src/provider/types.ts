import { Opaque } from 'type-fest';

export interface GitLabUser {
  id: number;
  name: string;
  username: string;
  created_at: string;
  web_url: string;
  state: string;
  email: string;
  public_email: string;
  is_admin: boolean;
  can_create_group: boolean;
  can_create_project: boolean;
  two_factor_enabled: boolean;
  external: boolean;
  private_profile: boolean;
  trial: boolean;
}

export interface GitLabUserRef {
  id: number;
  name: string;
  username: string;
  state: string;
  access_level: number;
}

export interface GitLabGroup {
  id: number;
  parent_id: number | null;
  name: string;
  web_url: string;
  created_at: string;
  path: string;
  description: string;
  visibility: string;
  share_with_group_lock: boolean;
  require_two_factor_authentication: boolean;
  two_factor_grace_period: number;
  project_creation_level: string;
  auto_devops_enabled: boolean;
  subgroup_creation_level: string;
  emails_disabled: boolean;
  mentions_disabled: boolean;
  lfs_enabled: boolean;
  default_branch_protection: number;
  request_access_enabled: boolean;
  full_name: string;
  full_path: string;
  file_template_project_id: number;
}

export interface GitLabFinding {
  id: number;
  report_type?: string;
  name?: string;
  severity?: string;
  confidence?: string;
  scanner?: {
    external_id?: string;
    name?: string;
    vendor?: string;
  };
  identifiers?: [
    {
      external_type: string;
      external_id: string;
      name: string;
      url: string;
    },
  ];
  project_fingerprint?: string;
  uuid?: string;
  create_jira_issue_url?: string;
  false_positive?: boolean;
  create_vulnerability_feedback_issue_path?: string;
  create_vulnerability_feedback_merge_request_path?: string;
  create_vulnerability_feedback_dismissal_path?: string;
  project?: {
    id?: number;
    name?: string;
    full_path?: string;
    full_name?: string;
  };
  dismissal_feedback?: string;
  issue_feedback?: string;
  merge_request_feedback?: string;
  description?: null;
  links?: [{ url: string }];
  location?: {
    file?: string;
    start_line?: number;
    class?: string;
    method?: string;
    hostname?: string;
  };
  solution?: string;
  evidence?: string;
  details?: {
    urls: { name?: string; type?: string };
    discovered_at?: {
      name?: string;
      type?: string;
      value?: string;
    };
  };
  state?: string;
  scan?: {
    type?: string;
    status?: string;
    start_time?: string;
    end_time?: string;
  };
  blob_path?: string;
}

export type GitLabVersion = {
  version: string;
  revision: string;
  enterprise: boolean;
};

export type GitLabProject = Opaque<any, 'GitLabProject'>;

export type GitLabMergeRequest = {
  id: number;
  iid: number;
  project_id: number;
  title: string;
  description: string;
  state: string;
  created_at: string;
  updated_at: string;
  /** @deprecated Deprecated in GitLab 14.7, and scheduled for removal in API version 5. Use merge_user instead. */
  merged_by?: Author;
  merge_user: Author;
  merged_at: string;
  closed_by: any;
  closed_at: any;
  target_branch: string;
  source_branch: string;
  user_notes_count: number;
  upvotes: number;
  downvotes: number;
  author: Author;
  assignees: any[];
  assignee: any;
  reviewers: any[];
  source_project_id: number;
  target_project_id: number;
  labels: any[];
  draft: boolean;
  imported: boolean;
  imported_from: string;
  /** @deprecated Use draft instead. */
  work_in_progress?: boolean;
  milestone: any;
  merge_when_pipeline_succeeds: boolean;
  /** @deprecated Deprecated in GitLab 15.6. Use detailed_merge_status instead. */
  merge_status?: string;
  detailed_merge_status: string;
  sha: string;
  merge_commit_sha: string;
  squash_commit_sha: any;
  discussion_locked: any;
  should_remove_source_branch: boolean;
  force_remove_source_branch: boolean;
  prepared_at: string;
  /** @deprecated Deprecated in GitLab 12.7, and scheduled for removal in API version 5. Use references instead. */
  reference?: string;
  references: References;
  web_url: string;
  time_stats: TimeStats;
  squash: boolean;
  squash_on_merge: boolean;
  task_completion_status: TaskCompletionStatus;
  has_conflicts: boolean;
  blocking_discussions_resolved: boolean;
  /** @deprecated Deprecated in GitLab 16.0. Premium and Ultimate only. */
  approvals_before_merge?: any;

  // TODO: check if these are actually in the response, keeping them here for now.
  allow_collaboration: boolean;
  allow_maintainer_to_push: boolean;
};

interface Author {
  id: number;
  username: string;
  name: string;
  state: string;
  locked: boolean;
  avatar_url: string;
  web_url: string;
}

interface References {
  short: string;
  relative: string;
  full: string;
}

interface TimeStats {
  time_estimate: number;
  total_time_spent: number;
  human_time_estimate: any;
  human_total_time_spent: any;
}

interface TaskCompletionStatus {
  count: number;
  completed_count: number;
}

export type GitLabMergeCommitRequest = Opaque<any, 'GitLabMergeCommitRequest'>;

// https://docs.gitlab.com/ee/api/merge_request_approvals.html#get-configuration-1
export type GitLabMergeRequestApproval = Opaque<
  any,
  'GitLabMergeRequestApproval'
>;

export interface GitlabLabel {
  id: number;
  color: string;
  text_color: string;
  name: string;
  description?: string | null;
  description_html?: string | null;
  open_issues_count: number;
  closed_issues_count: number;
  open_merge_requests_count: number;
  subscribed: boolean;
  priority?: number;
  is_project_label?: boolean | null;
}
