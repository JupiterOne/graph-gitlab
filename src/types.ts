export interface GitlabIntegrationConfig {
  baseUrl: string;
  personalToken: string;

  /**
   * Used to limit number of ingested merge requests.
   */
  mergeRequestsUpdatedAfter?: Date;
}
