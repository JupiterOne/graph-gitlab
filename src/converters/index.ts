import {
  createIntegrationEntity,
  Entity,
  parseTimePropertyValue,
} from '@jupiterone/integration-sdk-core';

import { Entities } from '../constants';
import {
  GitLabMergeRequest,
  GitLabMergeCommitRequest,
  GitLabMergeRequestApproval,
  GitLabProject,
  GitLabUser,
  GitLabFinding,
  GitlabLabel,
  GitLabVersion,
} from '../provider/types';
import { getCommitWebLinkFromMergeRequest } from '../util/mergeRequest';

export function createAccountEntity(
  tokenOwner: GitLabUser,
  systemVersion: GitLabVersion,
): Entity {
  return createIntegrationEntity({
    entityData: {
      source: { systemVersion, tokenOwner },
      assign: {
        _key: createAccountEntityKey(systemVersion.version),
        _type: Entities.ACCOUNT._type,
        _class: Entities.ACCOUNT._class,
        id: tokenOwner.id.toString(),
        name: tokenOwner.name,
        version: systemVersion.version,
        revision: systemVersion.revision,
        enterprise: systemVersion.enterprise,
      },
    },
  });
}

const ACCOUNT_VERSION_PREFIX = 'gitlab-account-version';
export function createAccountEntityKey(version: string): string {
  return `${ACCOUNT_VERSION_PREFIX}:${version}`;
}

export function createProjectEntity(project: GitLabProject): Entity {
  const { _links, ...source } = project; // drop the _links
  const key = createProjectEntityIdentifier(project.id);

  return createIntegrationEntity({
    entityData: {
      source,
      assign: {
        _key: key,
        _type: Entities.PROJECT._type,
        _class: Entities.PROJECT._class,

        id: project.id.toString(),
        name: project.name,
        owner: project.owner?.name,
        createdOn: new Date(project.created_at).getTime(),
        description: project.description || undefined,
        webLink: project.web_url,
        visibility: project.visibility,
        public: project.visibility !== 'private',
        issuesEnabled: project.issues_enabled,
        mergeRequestsEnabled: project.merge_requests_enabled,
        jobsEnabled: project.jobs_enabled,
        wikiEnabled: project.wiki_enabled,
        snippetsEnabled: project.snippets_enabled,
        canCreateMergeRequestIn: project.can_create_merge_request_in,
        resolveOutdatedDiffDiscussions:
          project.resolve_outdated_diff_discussions,
        containerRegistryEnabled: project.container_registry_enabled,
        archived: project.archived,
        sharedRunnersEnabled: project.shared_runners_enabled,
        publicJobs: project.public_jobs,
        allowMergeOnSkippedPipeline: !!project.allow_merge_on_skipped_pipeline,
        onlyAllowMergeIfPipelineSucceeds:
          project.only_allow_merge_if_pipeline_succeeds,
        onlyAllowMergeIfAllDiscussionsAreResolved:
          project.only_allow_merge_if_all_discussions_are_resolved,
        removeSourceBranchAfterMerge: project.remove_source_branch_after_merge,
        requestAccessEnabled: project.request_access_enabled,
        autocloseReferencedIssues: project.autoclose_referenced_issues,
        topics: project.topics,
        fullName: project.path_with_namespace,
      },
    },
  });
}

const PROJECT_ID_PREFIX = 'gitlab-project';
export function createProjectEntityIdentifier(id: number): string {
  return `${PROJECT_ID_PREFIX}:${id}`;
}

export function createUserEntity(user: GitLabUser): Entity {
  const key = createUserEntityIdentifier(user.id);

  return createIntegrationEntity({
    entityData: {
      source: user,
      assign: {
        _key: key,
        _type: Entities.USER._type,
        _class: Entities.USER._class,
        id: user.id.toString(),
        name: user.name,
        createdOn: new Date(user.created_at).getTime(),
        webLink: user.web_url,
        username: user.username,
        state: user.state,
        email: user.email || user.public_email || undefined,
        publicEmail: user.public_email,
        isAdmin: user.is_admin,
        canCreateGroup: user.can_create_group,
        canCreateProject: user.can_create_project,
        twoFactorEnabled: user.two_factor_enabled,
        external: user.external,
        privateProfile: user.private_profile,
        trial: user.trial,
        active: user.state === 'active',
      },
    },
  });
}

const USER_ID_PREFIX = 'gitlab-user';

export function createUserEntityIdentifier(id: number): string {
  return `${USER_ID_PREFIX}:${id}`;
}

export function createMergeRequestEntity(
  projectName: string,
  mergeRequest: GitLabMergeRequest,
  approval?: GitLabMergeRequestApproval,
): Entity {
  const key = createMergeRequestEntityIdentifier(mergeRequest.id);

  return createIntegrationEntity({
    entityData: {
      source: mergeRequest,
      assign: {
        _key: key,
        _type: Entities.MERGE_REQUEST._type,
        _class: Entities.MERGE_REQUEST._class,
        _rawData: [
          {
            name: 'approval',
            rawData: approval,
          },
        ],

        id: String(mergeRequest.id),
        iid: mergeRequest.iid,
        projectId: mergeRequest.project_id,
        name: mergeRequest.title,
        title: mergeRequest.title,
        state: mergeRequest.state,
        source: mergeRequest.source_branch,
        target: mergeRequest.target_branch,
        repository: projectName,
        createdOn: parseTimePropertyValue(mergeRequest.created_at),
        updatedOn: parseTimePropertyValue(mergeRequest.updated_at),
        authorId: mergeRequest.author.id,
        webLink: mergeRequest.web_url,
        mergeWhenPipelineSucceeds: mergeRequest.merge_when_pipeline_succeeds,
        shouldRemoveSourceBranch: !!mergeRequest.should_remove_source_branch,
        forceRemoveSourceBranch: mergeRequest.force_remove_source_branch,
        allowCollaboration: mergeRequest.allow_collaboration,
        allowMaintainerToPush: mergeRequest.allow_maintainer_to_push,
        squash: mergeRequest.squash,
        approved: approval?.approved,
        approvers: approval?.approved_by?.map((approver) => approver.user.name),
        approverIds: approval?.approved_by?.map((approver) =>
          String(approver.user.id),
        ),
        approverLogins: approval?.approved_by?.map(
          (approver) => approver.user.username,
        ),
        mergedOn: parseTimePropertyValue(mergeRequest.merged_at),
        closedOn: parseTimePropertyValue(mergeRequest.closed_at),
        sha: mergeRequest.sha,
        mergeCommitSha: mergeRequest.merge_commit_sha || undefined,
        squashCommitSha: mergeRequest.squash_commit_sha || undefined,
        commitWebLink: getCommitWebLinkFromMergeRequest(mergeRequest),
      },
    },
  });
}

export function createMergeRequestCommitEntity(
  mergeRequestCommit: GitLabMergeCommitRequest,
): Entity {
  const key = createCommitIdentifier(mergeRequestCommit.id);

  return createIntegrationEntity({
    entityData: {
      source: mergeRequestCommit,
      assign: {
        _key: key,
        _type: Entities.COMMIT._type,
        _class: Entities.COMMIT._class,
        id: String(mergeRequestCommit.id),
        shortId: String(mergeRequestCommit.short_id),
        title: mergeRequestCommit.title,
        name: mergeRequestCommit.title,
        displayName: mergeRequestCommit.title,
        merge: false,
        versionBump: false,
        createdOn: parseTimePropertyValue(mergeRequestCommit.created_at),
        webLink: mergeRequestCommit.web_url,
        message: mergeRequestCommit.message,
        authoredOn: parseTimePropertyValue(mergeRequestCommit.authored_date),
        committedOn: parseTimePropertyValue(mergeRequestCommit.committed_date),
        commitWebLink: mergeRequestCommit.web_url,
        committerName: mergeRequestCommit.committer_name,
        committerEmail: mergeRequestCommit.committer_email,
        authorName: mergeRequestCommit.author_name,
        authorEmail: mergeRequestCommit.author_email,
      },
    },
  });
}

const MERGE_REQUEST_ID_PREFIX = 'gitlab-merge-request';
export function createMergeRequestEntityIdentifier(id: number): string {
  return `${MERGE_REQUEST_ID_PREFIX}:${id}`;
}

const COMMIT_ID_PREFIX = 'gitlab-commit';
export function createCommitIdentifier(id: number): string {
  return `${COMMIT_ID_PREFIX}:${id}`;
}

export function createVulnerabilityFindingEntity(
  finding: GitLabFinding,
): Entity {
  const key = createFindingIdentifier(finding.id);

  // For more Finding data model class required fields info please visit: https://github.com/JupiterOne/data-model/blob/main/src/schemas/Finding.json.
  return createIntegrationEntity({
    entityData: {
      source: finding,
      assign: {
        _key: key,
        _type: Entities.FINDING._type,
        _class: Entities.FINDING._class,
        // START: finding data model required fields
        category: finding?.report_type,
        severity: finding?.severity,
        numericSeverity: getNumericSeverity(finding?.severity),
        open: finding.state?.toLocaleLowerCase() === 'detected',
        // END: finding data model required fields
        id: String(finding.id),
        reportType: finding.report_type,
        name: finding.name,
        confidence: finding.confidence,
        'scanner.externalId': finding.scanner?.external_id,
        'scanner.name': finding.scanner?.name,
        'scanner.vendor': finding.scanner?.vendor,
        identifiers: finding.identifiers?.map(
          (identifier) => identifier.external_id,
        ),
        projectFingerprint: finding.project_fingerprint,
        uuid: finding.uuid,
        createJiraIssueUrl: finding.create_jira_issue_url,
        falsePositive: finding.false_positive,
        createVulnerabilityFeedbackIssuePath:
          finding.create_vulnerability_feedback_issue_path,
        createVulnerabilityFeedbackMergeRequestPath:
          finding.create_vulnerability_feedback_merge_request_path,
        createVulnerabilityFeedbackDismissalPath:
          finding.create_vulnerability_feedback_dismissal_path,
        'project.id': finding.project?.id,
        'project.name': finding.project?.name,
        'project.fullPath': finding.project?.full_path,
        'project.fullName': finding.project?.full_name,
        dismissalFeedback: finding.dismissal_feedback,
        issueFeedback: finding.issue_feedback,
        mergeRequestFeedback: finding.merge_request_feedback,
        description: finding.description,
        links: finding.links?.map((link) => link.url),
        'location.file': finding.location?.file,
        'location.startLine': finding.location?.start_line,
        'location.class': finding.location?.class,
        'location.method': finding.location?.method,
        'location.hostname': finding.location?.hostname,
        solution: finding.solution,
        evidence: finding.evidence,
        state: finding.state,
        blobPath: finding.blob_path,
        'scan.type': finding.scan?.type,
        'scan.status': finding.scan?.status,
        'scan.startTime': parseTimePropertyValue(finding.scan?.start_time),
        'scan.endTime': parseTimePropertyValue(finding.scan?.end_time),
      },
    },
  });
}

const FINDING_ID_PREFIX = 'gitlab-finding';
export function createFindingIdentifier(id: number): string {
  return `${FINDING_ID_PREFIX}:${id}`;
}

export function getNumericSeverity(severity: string | undefined) {
  if (!severity) {
    return 0;
  } else if (/critical/i.test(severity)) {
    return 10;
  } else if (/high/i.test(severity)) {
    return 7;
  } else if (/medium/i.test(severity)) {
    return 5;
  } else if (/low/i.test(severity)) {
    return 2;
  } else if (/info/i.test(severity)) {
    return 1;
  } else if (/unknown/i.test(severity)) {
    return undefined;
  }
}

export function createLabelEntity(label: GitlabLabel): Entity {
  const key = createLabelEntityIdentifier(label.id);

  return createIntegrationEntity({
    entityData: {
      source: label,
      assign: {
        _key: key,
        _type: Entities.LABEL._type,
        _class: Entities.LABEL._class,
        id: label.id.toString(),
        displayName: label.name,
        name: label.name,
        description: label.description ?? undefined,
        color: label.color,
        textColor: label.text_color,
        openIssuesCount: label.open_issues_count,
        closedIssuesCount: label.closed_issues_count,
        openMergeRequestsCount: label.open_merge_requests_count,
        subscribed: label.subscribed,
        priority: label.priority,
      },
    },
  });
}

const LABEL_ID_PREFIX = 'gitlab-label';

export function createLabelEntityIdentifier(id: number): string {
  return `${LABEL_ID_PREFIX}:${id}`;
}
