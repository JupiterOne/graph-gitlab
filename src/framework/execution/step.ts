import uniq from 'lodash/uniq';

import {
  buildStepDependencyGraph,
  executeStepDependencyGraph,
} from './dependencyGraph';

import {
  IntegrationStep,
  IntegrationExecutionContext,
  IntegrationStepResult,
  IntegrationStepResultStatus,
  IntegrationStepStartStates,
  PartialDatasets,
} from './types';

export async function executeSteps(
  context: IntegrationExecutionContext,
  steps: IntegrationStep[],
  stepStartStates: IntegrationStepStartStates,
): Promise<IntegrationStepResult[]> {
  const stepGraph = buildStepDependencyGraph(steps);
  return executeStepDependencyGraph(context, stepGraph, stepStartStates);
}

export function getDefaultStepStartStates(
  steps: IntegrationStep[],
): IntegrationStepStartStates {
  return steps.reduce(
    (states: IntegrationStepStartStates, step: IntegrationStep) => {
      states[step.id] = {
        disabled: false,
      };
      return states;
    },
    {},
  );
}

export function determinePartialDatasetsFromStepExecutionResults(
  stepResults: IntegrationStepResult[],
): PartialDatasets {
  return stepResults.reduce(
    (partialDatasets: PartialDatasets, stepResult: IntegrationStepResult) => {
      if (stepResult.status !== IntegrationStepResultStatus.SUCCESS) {
        partialDatasets.types = uniq(
          partialDatasets.types.concat(stepResult.types),
        );
      }
      return partialDatasets;
    },
    { types: [] },
  );
}
