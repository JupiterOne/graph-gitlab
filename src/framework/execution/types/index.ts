export * from './context';
export * from './step';
export * from './config';
export * from './validation';
export * from './instance';
export * from './logger';
export * from './partialDatasets';
export * from './jobState';
