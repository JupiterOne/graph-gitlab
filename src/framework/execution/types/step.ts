import { IntegrationStepExecutionContext } from './context';

export enum IntegrationStepResultStatus {
  SUCCESS = 'success',
  FAILURE = 'failure',
  PARTIAL_SUCCESS_DUE_TO_DEPENDENCY_FAILURE = 'partial_success_due_to_dependency_failure',
  SKIPPED = 'skipped',
  NOT_EXECUTED = 'not_executed',
}

export type IntegrationStep = IntegrationStepMetadata & {
  /**
   * Function that runs to perform the stpe that
   */
  executionHandler: (
    context: IntegrationStepExecutionContext,
  ) => Promise<void> | void;
};

export type IntegrationStepResult = IntegrationStepMetadata & {
  status: IntegrationStepResultStatus;
};

interface IntegrationStepMetadata {
  /*
   * Identifier used to reference and track steps
   */
  id: string;

  /**
   * Friendly name that will be displayed in debug logs
   * and to customers in the job event log.
   */
  name: string;

  /**
   * Entity or relationship types that are expected to be
   * generated by this step
   */
  types: string[];

  /**
   * An optional array of other step ids that need to execute
   * before the current step can.
   */
  dependsOn?: string[];
}
