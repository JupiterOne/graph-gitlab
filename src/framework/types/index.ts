export * from './entity';
export * from './relationship';
