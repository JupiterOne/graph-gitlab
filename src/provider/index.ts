import { IntegrationInstance } from '@jupiterone/integration-sdk-core';

import { GitlabIntegrationConfig } from '../types';
import { GitlabClient } from './GitlabClient';

type GitlabClientConfig = {
  baseUrl: string;
  personalToken: string;
};

export function createGitlabClient(config: GitlabClientConfig): GitlabClient {
  return new GitlabClient(config.baseUrl, config.personalToken);
}

export type ClientCreator = (
  client: IntegrationInstance<GitlabIntegrationConfig>,
) => GitlabClient;
