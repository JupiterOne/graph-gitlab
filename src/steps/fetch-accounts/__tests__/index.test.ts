import {
  createMockIntegrationLogger,
  createMockStepExecutionContext,
} from '@jupiterone/integration-sdk-testing';

import step from '../';
import { Recording, setupRecording } from '../../../../test';
import { createGitlabClient } from '../../../provider';
import { GitLabUser } from '../../../provider/types';
import { createAccountEntity } from '../../../converters';

let recording: Recording;

beforeEach(() => {
  recording = setupRecording({
    directory: __dirname,
    name: 'accounts',
  });
});

afterEach(async () => {
  await recording.stop();
});

test('Account fetching', async () => {
  const context = createMockStepExecutionContext({
    instanceConfig: {
      baseUrl: process.env.BASE_URL || 'https://gitlab.com',
      personalToken: process.env.PERSONAL_TOKEN || 'string-value',
    },
  });
  const provider = createGitlabClient(
    context.instance.config,
    createMockIntegrationLogger(),
  );

  const results = await provider.fetchTokenOwner();

  expect(results).toEqual(
    expect.objectContaining({
      id: expect.any(Number),
      name: expect.any(String),
    }),
  );
});

test('Account entity conversion', () => {
  const tokenOwner = {
    id: 1,
    username: 'account',
    name: 'account',
    created_at: '2020-01-01T00:00:00.000Z',
  } as GitLabUser;

  const systemVersion = {
    version: 'testv1',
    revision: '234',
    enterprise: false,
  };

  const entity = createAccountEntity(tokenOwner, systemVersion);

  expect(entity).toEqual(
    expect.objectContaining({
      _key: 'gitlab-account-version:testv1',
      _type: 'gitlab_account',
      _class: ['Account'],
      id: '1',
      name: 'account',
      displayName: 'account',
      enterprise: false,
      revision: '234',
      version: 'testv1',
      _rawData: [
        {
          name: 'default',
          rawData: { tokenOwner, systemVersion },
        },
      ],
    }),
  );
});

test('step data collection', async () => {
  const context = createMockStepExecutionContext({
    instanceConfig: {
      baseUrl: process.env.BASE_URL || 'https://gitlab.com',
      personalToken: process.env.PERSONAL_TOKEN || 'string-value',
    },
  });
  await step.executionHandler(context);

  expect(context.jobState.collectedEntities).toHaveLength(1);
  expect(context.jobState.collectedRelationships).toHaveLength(0);
  expect(context.jobState.collectedEntities).toMatchSnapshot();
});
