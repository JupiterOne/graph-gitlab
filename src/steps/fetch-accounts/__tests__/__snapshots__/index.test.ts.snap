// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`step data collection 1`] = `
Array [
  Object {
    "_class": Array [
      "Account",
    ],
    "_key": "gitlab-account-version:17.0.0-pre",
    "_rawData": Array [
      Object {
        "name": "default",
        "rawData": Object {
          "systemVersion": Object {
            "enterprise": true,
            "kas": Object {
              "enabled": true,
              "externalUrl": "wss://kas.gitlab.com",
              "version": "17.0.0-rc4",
            },
            "revision": "41e5f397063",
            "version": "17.0.0-pre",
          },
          "tokenOwner": Object {
            "avatar_url": "https://secure.gravatar.com/avatar/547145a05e7cecb80611baf055925182?s=80&d=identicon",
            "bio": null,
            "can_create_group": true,
            "can_create_project": true,
            "color_scheme_id": 1,
            "confirmed_at": "2020-04-28T12:33:25.235Z",
            "created_at": "2020-04-28T12:33:25.275Z",
            "current_sign_in_at": "2020-05-08T17:05:49.796Z",
            "email": "codeworkr@gmail.com",
            "external": false,
            "extra_shared_runners_minutes_limit": null,
            "id": 5887285,
            "identities": Array [
              Object {
                "extern_uid": "30329077",
                "provider": "github",
                "saml_provider_id": null,
              },
            ],
            "job_title": "",
            "last_activity_on": "2020-05-12",
            "last_sign_in_at": "2020-05-08T11:19:04.307Z",
            "linkedin": "",
            "location": null,
            "name": "eXtremeX",
            "organization": null,
            "private_profile": false,
            "projects_limit": 100000,
            "public_email": "",
            "shared_runners_minutes_limit": null,
            "skype": "",
            "state": "active",
            "theme_id": 1,
            "twitter": "",
            "two_factor_enabled": false,
            "username": "eXtremeXR",
            "web_url": "https://gitlab.com/eXtremeXR",
            "website_url": "",
            "work_information": null,
          },
        },
      },
    ],
    "_type": "gitlab_account",
    "createdOn": undefined,
    "displayName": "eXtremeX",
    "enterprise": true,
    "id": "5887285",
    "name": "eXtremeX",
    "revision": "41e5f397063",
    "version": "17.0.0-pre",
  },
]
`;
